src/plugins/vis_type_table_new/public/utils/convert_to_formatted_data.ts

```diff
@@ -0,0 +1,179 @@
+  return newArray;
+}
+
+/**
+ * @param columns - the formatted columns that will be displayed
+ * @param title - the title of the column to add to
+ * @param rows - the row data for the columns
+ * @param insertAtIndex - the index to insert the percentage column at
+ * @returns cols and rows for the table to render now included percentage column(s)
+ */
+function addPercentageCol(
+  columns: FormattedColumn[],
+  title: string,
+  rows: Table['rows'],
+  insertAtIndex: number
+) {
+  const { id, sumTotal } = columns[insertAtIndex];
+  const newId = `${id}-percents`;
+  const formatter = getFormatService().deserialize({ id: 'percent' });
+  const i18nTitle = i18n.translate('visTypeTableNew.params.percentageTableColumnName', {
+    defaultMessage: '{title} percentages',
+    values: { title },
+  });
+  const newCols = insert(columns, insertAtIndex, {
+    title: i18nTitle,
+    id: newId,
+    formatter,
+    filterable: false,
+  });
+  const newRows = rows.map((row) => ({
+    [newId]: (row[id] as number) / (sumTotal as number),
+    ...row,
+  }));
+
+  return { cols: newCols, rows: newRows };
+}
+
+export interface FormattedDataProps {
+  formattedRows: OpenSearchDashboardsDatatableRow[];
+  formattedColumns: FormattedColumn[];
+}
+
+export const convertToFormattedData = (
+  table: Table,
+  visConfig: TableVisConfig
+): FormattedDataProps => {
+  const { buckets, metrics } = visConfig;
+  let formattedRows: OpenSearchDashboardsDatatableRow[] = table.rows;
+  let formattedColumns: FormattedColumn[] = table.columns
+    .map(function (col, i) {
+      const isBucket = buckets.find((bucket) => bucket.accessor === i);
+      const dimension = isBucket || metrics.find((metric) => metric.accessor === i);
+
+      if (!dimension) return undefined;
+
+      const formatter = getFormatService().deserialize(dimension.format);
+
+      const formattedColumn: FormattedColumn = {
+        id: col.id,
+        title: col.name,
+        formatter,
+        filterable: !!isBucket,
+      };
+
+      const isDate = dimension?.format?.id === 'date' || dimension?.format?.params?.id === 'date';
+      const allowsNumericalAggregations = formatter?.allowsNumericalAggregations;
+
+      if (allowsNumericalAggregations || isDate || visConfig.totalFunc === AggTypes.COUNT) {
+        const sum = table.rows.reduce((prev, curr) => {
+          // some metrics return undefined for some of the values
+          // derivative is an example of this as it returns undefined in the first row
+          if (curr[col.id] === undefined) return prev;
+          return prev + (curr[col.id] as number);
+        }, 0);
+
+        formattedColumn.sumTotal = sum;
+        switch (visConfig.totalFunc) {
+          case AggTypes.SUM: {
+            if (!isDate) {
+              formattedColumn.formattedTotal = formatter?.convert(sum);
+              formattedColumn.total = formattedColumn.sumTotal;
+            }
+            break;
+          }
```

Comment:
The whole idea of a single "total function" is not really valid, and something we'll want to revisit (for example, summing averages or medians is just nonsense). I'll open a follow-up issue.

Reply:
Okay. Thanks. Could you also add the `tableVis` tag in the follow-up issue, so I could track it?

Reply:
yeah, will do